README.md
# Ninmonkey.Console

Tools for a better console experience

- [Ninmonkey.Console](#ninmonkeyconsole)
- [Useful Commands](#useful-commands)
  - [`Fzf` : Select Values like a simplified `Out-GridView -PassThru`](#fzf--select-values-like-a-simplified-out-gridview--passthru)
  - [Text output](#text-output)
  - [Conversions](#conversions)
- [Custom Verbs](#custom-verbs)
- [Frequently used](#frequently-used)
- [Console Encoding](#console-encoding)
- [AppxPackages : Windows Apps](#appxpackages--windows-apps)
- [Types](#types)
  - [Format-TypeName](#format-typename)
  - [Format-GenericTypeName](#format-generictypename)
  - [Test-NullArg](#test-nullarg)
- [See More](#see-more)
  - [libs](#libs)
  - [vexx32.github.io : Autocomplete and more](#vexx32githubio--autocomplete-and-more)
  - [Docs](#docs)

# Useful Commands


| Name           | Description                                 |
| -------------- | ------------------------------------------- |
| `Hex`          | `ConvertTo-HexString`                       |
| `Bits`         | `ConvertTo-BitString`                       |
| `Base64`       | .                                           |
| `Get-EnumInfo` | .                                           |
| `Out-Fzf`      | .                                           |
| `Out-Code`     | .                                           |
| `Goto`         | `cd` to directory, works on non-directories |
| `Here`         | Open filemanager                            |

## `Fzf` : Select Values like a simplified `Out-GridView -PassThru`

```powershell
🐒> Get-ChildItem $Env:APPDATA -Directory | Out-Fzf | Format-RelativePath $Env:APPDATA
.\Code
.\Code - Insiders
```
```powershell
🐒> $Fzf
C:\Users\monkey\AppData\Roaming\Code
C:\Users\monkey\AppData\Roaming\Code - Insiders
# $Fzf was set because it defaults to:
$PSDefaultParameterValues['Out-Fzf:OutVariable'] = 'Fzf'
```

## Text output
| Name    | Description                  |
| ------- | ---------------------------- |
| `Label` | Colored `Label: Value` pairs |
| `H1`    | Heading text                 |
| `Br`    | Add newlines                 |
| `Hr`    | Horizontal Rule with padding |


## Conversions

```powershell
🐒> $numbers = 0, 127, 0xffef
  >> $numbers | Hex
0x0
0x7f
0xffef

🐒> $numbers | Bits
0000.0000
0111.1111
1111.1111 1110.1111

🐒> '0x34' | Number # returned type [Int32]

🐒> $bytes | Base64
```

# Custom Verbs

| Name                  | Description                                                                                                                                                                                                                                                                                                                        |
| --------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `Ensure->`            | ensure condition is true, else make it so, <br>like: `Ensure-FileExists`, `Ensure-CurrentDirIs`, etc.<br>The intent is non-fatal asserts on a state, which are silently resolved -- if needed                                                                                                                                      |
| `Resolve->`           | ConvertType to another type -- if needed.<Br>Like: `Resolve-TypeInfo`, ..<br/>`Resolve->FileInfo` attempts to return a `Get-Item`. If it fails, return the original string, instead of errors or nothing. <br>`Resolve->` Is similar to `Ensure->`, the difference is `Resolve->` returns objects . Ensure is about a valid state. |
| `To->`  and  `From->` | Sugar for `ConvertTo-Type` and `ConvertFrom-Type` / load from                                                                                                                                                                                                                                                                      |
| `Fmt->`               | Custom Formatting of objects and strings depending on the command<br>Built to chain off of one another, unlike `Format-*` functions                                                                                                                                                                                                |

# Frequently used

```ps1
# Jump to modules
🐒> Get-Module PSKoans | Goto 
    pwd

C:\Users\monkey\Documents\PowerShell\Modules\PSKoans\0.67.1
```

# Console Encoding

```powershell
🐒> Get-ConsoleEncoding

Name                      Encoding        CodePage isSingleByte
----                      --------        -------- ------------
OutputEncoding            Unicode (UTF-8)    65001        False
[console]::InputEncoding  Unicode (UTF-8)    65001        False
[console]::OutputEncoding Unicode (UTF-8)    65001        False
```

# AppxPackages : Windows Apps

Package names are not always descriptive. Sometimes nothing in the fields `Name`, `PackageName` or `FamilyName` are useful.

This example has the package name `Dayton` even though the game is named `State of Decay 2`
```powershell
🐒> Get-AppXPackage *state*decay*
# nothing found

🐒> $results = Get-NinAppXPackage 'state.*decay' -PassThru
🐒> $results | Format-List
```
```
Regex   : state.*decay
App     : Microsoft.Dayton_2.408.280.0_x64__8wekyb3d8bbwe
ExeList : {C:\Program Files\WindowsApps\Microsoft.Dayton_2.408.280.0_x64__8wekyb3d8bbwe\StateOfDecay2.exe, C:\Program Files\WindowsApps\Microsoft.Dayton_2.408.280.0_x64__8wekyb3d8bbwe\StateOfDecay2\Binaries\Win64\StateOfDecay2-Win64-Shipping.exe}
```
```powershell
🐒> Get-NinAppXPackage 'state.*decay'
```
```
Name              : Microsoft.Dayton
Publisher         : CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US
Architecture      : X64
ResourceId        :
Version           : 2.408.280.0
PackageFullName   : Microsoft.Dayton_2.408.280.0_x64__8wekyb3d8bbwe
InstallLocation   : C:\Program Files\WindowsApps\Microsoft.Dayton_2.408.280.0_x64__8wekyb3d8bbwe
IsFramework       : False
PackageFamilyName : Microsoft.Dayton_8wekyb3d8bbwe
PublisherId       : 8wekyb3d8bbwe
IsResourcePackage : False
IsBundle          : False
IsDevelopmentMode : False
NonRemovable      : False
Dependencies      : {Microsoft.VCLibs.140.00.UWPDesktop_14.0.29231.0_x64__8wekyb3d8bbwe}
IsPartiallyStaged : False
SignatureKind     : Store
Status            : Ok

Regex   : state.*decay
App     : Microsoft.Dayton_2.408.280.0_x64__8wekyb3d8bbwe
ExeList : {C:\Program Files\WindowsApps\Microsoft.Dayton_2.408.280.0_x64__8wekyb3d8bbwe\StateOfDecay2.exe, C:\Program Files\WindowsApps\Microsoft.Dayton_2.408.280.0_x64__8wekyb3d8bbwe\StateOfDecay2\Binaries\Win64\StateOfDecay2-Win64-Shipping.exe}

Regex        App
-----        ---
state.*decay Microsoft.Dayton_2.408.280.0_x64__8wekyb3d8bbwe
```

# Types


## Format-TypeName

```powershell
🐒> ls . | foreach-object{ $_.pstypenames } | sort -Unique
System.IO.DirectoryInfo
System.IO.FileInfo
System.IO.FileSystemInfo
System.MarshalByRefObject
System.Object

🐒>  ls . | foreach-object{ $_.pstypenames } | sort -Unique | Format-TypeName
[IO.DirectoryInfo]
[IO.FileInfo]
[IO.FileSystemInfo]
[MarshalByRefObject]
[Object]
```


## Format-GenericTypeName

```powershell
🐒> $items = [list[string]]::new()
  $items.GetType().FullName

System.Collections.Generic.List`1[[System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]

🐒> $items.GetType() | Format-TypeName

List`1[[String]]
```


## Test-NullArg


```powershell

🐒> 10, '', " ", $null, "`u{0}" | Test-NullArg | Format-Table

Value Type   IsNull IsNullOrWhiteSpace IsNullOrEmpty AsString ToString CastString TestId IsNullCodepoint
----- ----   ------ ------------------ ------------- -------- -------- ---------- ------ ---------------
   10 Int32   False              False         False '10'     '10'     '10'            0           False
      String  False               True          True ''       ''       ''              1            True
      String  False               True         False ' '      ' '      ' '             2           False
    ␀ [Null]   True               True          True ''       '␀'      ''              3           False
    ␀ String  False              False         False '␀'      '␀'      '␀'             4            True
```

# See More

## libs

- [ClassExplorer](https://github.com/seeminglyscience) by SeeminglyScience/ClassExplorer

## vexx32.github.io : Autocomplete and more

- [Creating Dynamic Sets for ValidateSet](https://vexx32.github.io/2018/11/29/Dynamic-ValidateSet/)
- [Working With Argument Transformations in PowerShell](https://vexx32.github.io/2018/12/13/Working-Argument-Transformations/)
- [Working with PowerShell's -replace Operator](https://vexx32.github.io/2019/03/20/PowerShell-Replace-Operator/)
- [ScriptBlocks and GetNewClosure\(\)](https://vexx32.github.io/2020/05/30/Scriptblock-GetNewClosure/)
- [Searching the PowerShell Abstract Syntax Tree](https://vexx32.github.io/2018/12/20/Searching-PowerShell-Abstract-Syntax-Tree/)
- [Implementing-ShouldProcess](https://vexx32.github.io/2018/11/22/Implementing-ShouldProcess/)
- [Invoke a PowerShell Module Command in the Global Scope](Invoke a PowerShell Module Command in the Global Scope)
  - > & (Get-Module 'Pester') { Get-Command -Module Pester }

## Docs
- Autocompletion [enum CompletionResultType ](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.completionresulttype?view=powershellsdk-7.0.0#System_Management_Automation_CompletionResultType)
- [IArgumentCompleter.CompleteArgument\(String, String, String, CommandAst, IDictionary\) Method](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.iargumentcompleter.completeargument?view=powershellsdk-7.0.0)


docs/changes.md
## Version 0.2.21

### Functions with `-Options` Parameters

- When a Function's parameter is named `-Options`, it takes a hashtable to modify behavior
- Internally, functions always reference `-Options` as the hashtable `$Config`. This is to make it clear it's the merged settings
- `-Options` functions often have a few templates that autocomplete. 

Settings are merged like this:
```ps1
$Config = mergeHashtable -BaseHash $DefaultValues -OtherHash $Options

# ex:
$Config = mergeHashtable -BaseHash @{ Name = 'fred' ; Region = 'east' } -OtherHash @{ Id = 4 ; Region = 'North' }
```

### Faster import

- Some things were cleaned up a bit, and I disabled some debug JSON files from writing on load.
- If import is printing any text on load, you can increase speed a bit by redirecting all streams to `$null`. (This applies to all modules).
```ps1
# silently import
Import-Module ninmonkey.Console -wa Ignore *>$null
```

- `Import-Module CompletionPredictor` is optional. 

It's not the default because it has a pretty significant import cost relative most modules.
```ps1
# my default
nin.ImportPSReadLine MyDefault_HistListView
# if you don't care about the initial import time, try the CompletionPredictor
nin.ImportPSReadLine Using_Plugin
```

### `Format-UnorderedList` aliased as `UL` and `join.UL`

- No longer errors when pipping nulls and empty strings to `Format-UnorderedList` without errors
- added a new alias `join.UL` for less ambiguity / consistency with future commands. `UL` is still defined. 

![image](img/format-ul-completer.png)

### New Experimental commands `Eye` and `io`

- You can use `Eye` to get a quick view of something. 
- For filterable output, `Inspect-ObjectProperty` aliased as `io` returns objects

```ps1
eye (get-date)
eye (gi .)
```
- `io` inspects and views `psobject.properties`
- `Eye` also uses `ClassExplorer\Find-Member`
```ps1
$PSStyle | io -SortBy Reported
| ? Type  -notmatch 'string'
| Ft Reported, Name, Type, IsNull, Is*, Value -AutoSize
```

examples

```ps1

$t = gi .

h1 'find non-blank properties'
$t 
| io -SortBy Reported
| ? IsBlank -Not
| Ft Reported, Name, Type, IsNull, Is*, Value -AutoSize

h1 'find blank properties'
$t
| io -sortBy Type
| ? IsBlank
| Ft Reported, Name, Type, IsNull, Is*, Value -AutoSize

h1 'ignore any null properties'
$t
| io -sortBy Type
| ? -not IsNull
| Ft Reported, Name, Type, IsNull, Is*, Value -AutoSize
```

![image](https://user-images.githubusercontent.com/3892031/201253572-be9547ec-4587-4521-a66d-d75c8f445782.png)
![image](https://user-images.githubusercontent.com/3892031/201253858-8091eb77-9747-4646-b4d7-ed711e7e3d10.png)

### bugfix

- Auto reset the log from `Enable-NinHistoryHandler`, if it when size >= 5MB.
- (Even before) It shouldn't have ran unless you explicitly call `Enable-NinHistoryHandler` 
- changed the message to a warning, to prevent accidentally enabling it in a profile

See: [changes.md](docs/changes.md) for more




root_autoloader.ps1
#Requires -Version 7.0

using namespace system.collections.generic
# __init__.ps1

Write-Warning "🐱‍👤finish $PSCommandPath"

$____debug = @{
    Export_Module_Listing_On_Startup = $false
    Export_Super_ExcessiveTest       = $false
}

# eaiser to manage and filter, especially a dynamic set, in one place
[hashtable]$script:publicToExport = @{
    'function' = @(
        'Join-Hashtable'
    )
    'alias'    = @(
        # when random errors occur, so far it's been PSScriptTools\Join-Hashtable causing exceptions
        # 'Join-Hashtable' # to be doubly sure to get priority? else put alias in my module
        # Putting in my profile
    )
    'cmdlet'   = @()
    'variable' = @()
    'meta'     = @()
    # 'formatData' = @()
}

$strUL = @{
    'sep' = "`n  - "
    'op'  = "`n  - "
    'os'  = "`n"
}


Write-Warning '[dev.nin] needs this'


Function Join-Hashtable {
    <#
    .description
        Copy and append BaseHash with new values from UpdateHash
    .notes
        # WIP
        future: add valuefrom pipeline to $UpdateHash param ?
    .example
        Join-Hashtable -Base @{ a = 1 }
    .link
        Ninmonkey.Console\Join-Hashtable
    .link
        Ninmonkey.Console\Merge-HashtableList
    .link
        Ninmonkey.Powershell\Join-Hashtable
    .link
        PSScriptTools\Join-Hashtable
    #>
    # [Alias('Join-HashTable')]
    [cmdletbinding()]
    [outputType('System.Collections.Hashtable')]
    param(
        # base hashtable
        [Alias('LeftHash')]
        [Parameter(Mandatory, Position = 0)]
        [hashtable]$BaseHash,

        # New values to append and/or overwrite
        [Parameter(Mandatory, Position = 1)]
        [hashtable]$OtherHash,

        # normal is to not modify left, return a new hashtable
        [Alias('UpdateOriginal')]
        [Parameter()][switch]$MutateLeft

        # default Left wins if they share a key name
        # [Parameter()][switch]$PrioritizeRight
    )

    # don't mutate $BaseHash
    process {
        if ($PrioritizeRight) {
            Throw 'Next to implement'
        }

        # $NewHash = [hashtable]::new( $BaseHash )
        if (! $MutateLeft ) {
            $TargetHash = [hashtable]::new( $BaseHash )
        }
        else {
            Write-Debug 'mutating left'
            $TargetHash = $BaseHash
        }

        # sometimes gives
        <#
             Method invocation failed because [System.Collections.Hashtable+KeyCollection] does not contain a method named 'Clone'.
        #>
        # $OtherHash.keys.clone() | ForEach-Object {
        $OtherHash.GetEnumerator() | ForEach-Object {
            $TargetHash[ $_ ] = $OtherHash[ $_ ]
        }
        return $TargetHash
    }
}


class ModuleExportRecord {
    <#
    One Idea is a pre-parse stage:
        [1] ast only, collect any metadata, and registered exports
        [2] potential filter on loaders
        [3] then actually invoke script blocks etc

    allow skipping function, parse without execute
    #>
    # future:// param that auto-logs when property is changed
    # maybe this is one instance per file
    # which gets appended to global state?
    [Collections.Generic.List[object]]$function = @()
    [Collections.Generic.List[object]]$alias = @()
    [Collections.Generic.List[object]]$cmdlet = @()
    [Collections.Generic.List[object]]$variable = @()
    [Collections.Generic.List[object]]$typeData = @()   # nyi
    [Collections.Generic.List[object]]$formatData = @() # nyi

    # store source filename etc?

    [Collections.Generic.List[object]]$meta = @() # nyi
}

# $ModuleExportRecord = [ModuleExportRecord]::new()

class NinModuleInfo {
    # global state
    #  PSRealine Exports
    # command line completers
    [Collections.Generic.List[ModuleExportRecord]]$ExportInfo = [Collections.Generic.List[ModuleExportRecord]]::new()

    [object]$PSReadlineHandlers = @() # or hooks
    [object]$NativeCommand = @()

    [void] AddExportInfo([ModuleExportRecord]$ExportInfo) {
        ## Add argument validation logic here
        # $this.Devices[$slot] = $dev
        $this.ExportInfo.Add( $ExportInfo )
    }
}

[ValidateNotNull()]
$script:ninModuleInfo = [NinModuleInfo]::new()


function Find-AutoloadChildItem {
    <#
    .synopsis
        filter out auto runner
    .example
        Find-AutoloadChildItem -path 'public_autoloader'
    #>
    [cmdletbinding()]
    param(
        # root dir to search. maybe allow paths piped
        [Alias('Path', 'PSPath')]
        [parameter(Mandatory)]
        [string[]]$InputPath,

        # names that will be converted to regex-literalsregex literals
        [parameter()]
        [string[]]$ignoreNamesLiteral = @(
            '.visual_tests.ps1',
            '.Interactive.ps1',
            '__init__.ps1',
            # '__init__.first.ps1'
            '.tests.ps1'
        ),
        # [parameter()]
        # [string[]]$ignoreNamesRegex = @()

        [hashtable]$Options = @{}
    )
    begin {
        $Config = @{
            RootFilter = '*.ps1'
            FilterFile = $true
            Recurse    = $false
        }
        # $Config = $Config | Join-Hashtable $Options

        $ignoreNamesLiteral = $ignoreNamesLiteral
        | ForEach-Object { [regex]::Escape($_) }
    }
    process {
        $curPath = $InputPath | Get-Item -ea stop
        $findAutoLoader_Splat = @{
            File    = $Config.FilterFile
            Path    = $curPath
            Filter  = $Config.RootFilter
            Recurse = $Config.Recurse
        }

        $findAutoLoader_Splat | Out-String | Write-Debug

        $filesQuery = Get-ChildItem @findAutoLoader_Splat


        $filteredQuery = $filesQuery
        | Where-Object {
            #future: filter as not directory? -File catches that
            $curFile = $_
            $match_tests = $ignoreNamesLiteral | ForEach-Object {
                $pattern = $_
                Write-Debug "test: '($curFile.Name)' -match '$pattern'"
                $curFile.Name -match $pattern
            }

            $answer = [bool](Test-Any $match_tests)
            # 'answer: ', $answer -join '' | Write-Information
            return -not $answer
        }

        if ($true) {
            # super gross, quick hack, because because Test-Any (where list) didn't seem to be matching correctly
            # prints: 'answer: False' 100% Of the time

            $doubleFilter = $filteredQuery
            | Where-Object { $_.Name -ne '__init__.ps1' }
            | Where-Object { $_.Name -ne '__init__.first.ps1' }
            | Where-Object { $_.Name -match '\.ps1$' }
            | Where-Object { $_.Name -notmatch '\.tests\.ps1$' }

            $filesQuery
            | Join-String @strUl -DoubleQuote
            | Join-String -op '1->filesQuery'
            | Out-String | Write-Debug
            # | Join-String -op 'BeforeFilter = ' -sep ', ' -DoubleQuote
            # | Out-String | Write-Debug

            $filteredQuery
            | Join-String @strUl -DoubleQuote
            | Join-String -op '2->filteredQuery'
            | Out-String | Write-Debug
            # | Join-String -op 'AfterFilter = ' -sep ', ' -DoubleQuote
            # | Out-String | Write-Debug

            $doubleFilter
            | Join-String @strUl -DoubleQuote
            | Join-String -op '3->DoubleFilter'
            | Out-String | Write-Debug
            return $doubleFilter
        }

    }
    end {

    }
}

# Export-ModuleMember -Function 'Find-AutoloadChildItem'
$NinModuleInfo.AddExportInfo(( [ModuleExportRecord]@{
            'function' = 'Join-Hashtable'
        }))


# $ModuleExportRecord.function.Add( 'Get-DirectChildItem' )

# [ModuleExportRecord]@{
#     'function' = 'Get-DirectChildItem'
# }

# $ninModuleInfo

# hardCoded until created
# see: <C:\Users\cppmo_000\Documents\2021\Powershell\My_Github\Dev.Nin\public_experiment\main_import_experimental.ps1>

# try {
# $fileList = @(
#     # 'Get-CommandSummary-OldMethod'
#     'Get-CommandSummafry'
#     'Find-Exception'
# )


# $x = $null
# return
function _find_moduleInitDir {
    <#
    .synopsis
        sugar to find '__init__.ps1' files, returning their parent directories
    .notes
        ask sci: how would I pass and invoke a custom sort order?
            _find_moduleInitDir -path . -Sort { $_.Name  } -Descending
    #>
    param(
        [Parameter()]
        [string]$Path = (Get-Item $PSScriptRoot)
    )
    $splat = @{
        Recurse = $true
        Force   = $true
        Path    = $Path
        Filter  = '__init__.ps1'
    }
    $dirsToLoad = Get-ChildItem @splat
    | ForEach-Object Directory
    # also test if eq to $PSScriptRoot?
    return $DirsToLoad
}



$dirsToLoad = _find_moduleInitDir
| Sort-Object { $_.FullName -match 'beforeAll' } -Descending  # future will allow sort order from '__init__.ps1'
$dirsToLoad | Write-Host -foreg Green
# |  % fullname

$DirsToLoad | Join-String @strUL | Write-Verbose
$DirsToLoad | Join-String @strUL { $_.Name } | Write-Debug
# | Write-Information
'before'

$NinModuleInfo | ConvertTo-Json -Depth 50 | Set-Content 'temp:\last_NinModuleInfo.json' -Encoding utf8
Get-Content 'temp:\last_NinModuleInfo.json' -Encoding utf8
| bat -l json -f -p | Write-Information


# $origTest = Get-ChildItem -File -Path (Get-Item -ea stop $PSScriptRoot)
$origTest = Get-ChildItem -File -Path $dirsToLoad
| Where-Object { $_.Name -ne '__init__.ps1' }
| Where-Object { $_.Name -ne '__init__.first.ps1' }
| Where-Object { $_.Name -match '\.ps1$' }
| Where-Object { $_.Name -notmatch '\.tests\.ps1$' }

$newTest = Find-AutoloadChildItem -InputPath $dirsToLoad -infa 'continue'
# | Sort-Object { $_.fullname -match 'zeroDepend_autoloader' } -Descending
| Sort-Object { $_.fullname -match '' } -Descending

$z = $Null
$z = $Null

class ParseResult {
    [IO.FileSystemInfo]$Name
    [bool]$Success
    [Collections.Generic.List[object]]$Error = [Collections.Generic.List[object]]::new()
}

[Collections.Generic.List[ParseResult]]$parseResultSummary = [Collections.Generic.List[ParseResult]]::new()

$newTest
| ForEach-Object {
    $Cur = $_
    $parseResult = [ParseResult]@{
        Name    = $Cur
        Success = $true
    }
    try {
        . $Cur
    }
    catch {
        $ParseResult.Success = $false
        $ParseResult.Error.AddRange( @($_) )
    }

    $parseResultSummary.Add( $parseResult )
}

# Don't dot tests, don't call self.
# Get-ChildItem -File -Path (Get-Item -ea stop $PSScriptRoot)
# | Where-Object { $_.Name -ne '__init__.ps1' }
# | Where-Object { $_.Name -ne '__init__.first.ps1' }
# | Where-Object { $_.Name -match '\.ps1$' }
# | Where-Object { $_.Name -notmatch '\.tests\.ps1$' }
# | ForEach-Object {
#     # $curScript = $_
#     # . $curScript
# }
# are these safe? or will it alter where-object?
# Write-Debug "[dev.nin] importing experiment '$($_.Name)'"
# } catch {
# Write-Error -Exception $_ -Message "DotsourceImportFailed: public_autoloader\__init__.ps1: '$($curScript)'" -TargetObject $_ -Category InvalidOperation
# Write-Error "todo: correctly throw: '$_'"
# }
# } catch {
# Write-Error "public_autoloader\__init__.ps1:  failed`ntodo: correctly throw: '$_'"
# Write-Error -Exception $_ -Message ''
# }

$script:publicToExport | Join-String -op 'ExperimentToExport' | Write-Debug

if ($ninModuleInfo.ExportInfo.function) {
    Export-ModuleMember -Function $ninModuleInfo.ExportInfo.function
}
if ($ninModuleInfo.ExportInfo.alias) {
    Export-ModuleMember -Alias $ninModuleInfo.ExportInfo.alias
}
if ($ninModuleInfo.ExportInfo.cmdlet) {
    Export-ModuleMember -Cmdlet $ninModuleInfo.ExportInfo.cmdlet
}
if ($ninModuleInfo.ExportInfo.variable) {
    Export-ModuleMember -Variable $ninModuleInfo.ExportInfo.variable
}
if ($ninModuleInfo.ExportInfo.typeData) {
    Write-Warning '$ninModuleInfo.TypeData export is NYI'
}
if ($ninModuleInfo.ExportInfo.formatData) {
    Write-Warning '$ninModuleInfo.FormatData export is NYI'
}

if ($script:publicToExport['function']) {
    Export-ModuleMember -Function $script:publicToExport['function']
}
if ($script:publicToExport['alias']) {
    Export-ModuleMember -Alias $script:publicToExport['alias']
}
if ($script:publicToExport['cmdlet']) {
    Export-ModuleMember -Cmdlet $script:publicToExport['cmdlet']
}
if ($script:publicToExport['variable']) {
    Export-ModuleMember -Variable $script:publicToExport['variable']
}

$meta | Format-Table | Out-String | Write-Information

if ($true) {
    # tip: 'Join-String' does not require the empty and $null gaurds, left in for intent
    # detect not yet transferred over
    if ($publicToExport.function.Count -gt 0) {
        $publicToExport.function | Join-String @strUl
        | Join-String -op "(eventually deprecated warning): `$publicToExport.function is not empty, migrate to 'ninModuleInfo' Items: `n"
        | Write-Debug
    }
    if ($publicToExport.alias.Count -gt 0) {
        $publicToExport.alias | Join-String @strUl
        | Join-String -op "(eventually deprecated warning): `$publicToExport.alias is not empty, migrate to 'ninModuleInfo' Items: `n"
        | Write-Debug

    }
    if ($publicToExport.cmdlet.Count -gt 0) {
        $publicToExport.cmdlet | Join-String @strUl | Write-Warning # | Write-Debug
        | Join-String -op "(eventually deprecated warning): `$publicToExport.cmdlet is not empty, migrate to 'ninModuleInfo' Items: `n"
        | Write-Debug
    }
    if ($publicToExport.variable.Count -gt 0) {
        $publicToExport.variable | Join-String @strUl | Write-Warning # | Write-Debug
        | Join-String -op "(eventually deprecated warning): `$publicToExport.variable is not empty, migrate to 'ninModuleInfo' Items: `n"
        | Write-Debug
    }
}
$devNinRoot = '~\.dev-nin\dump'


$__exportPaths = @{
    DevNinRoot                               = $devNinRoot
    LogExportBase                            = $devvNinRoot #'~\.dev-nin\dump' #'\console-last_import.json'
    Export_moduleExports_FromList            = Join-Path $devNinRoot 'console-last_list_import.json'
    Export_moduleExports_FromClass           = Join-Path $devNinRoot 'console-last_class_import.json'
    Export_moduleExports_FromClass_Expand    = Join-Path $devNinRoot 'console-last_class_import-ExportInfo.json'
    Export_moduleExports_parseResult         = Join-Path $devNinRoot 'console-last_parse_result.json'
    Export_moduleExports_parseResult_summary = Join-Path $devNinRoot 'console-last_parse_result-summary.json'

}

if ($true) {
}



if ($____debug.Export_Super_ExcessiveTest) {
    "`$____debug.Export_Super_ExcessiveTest: `$true from '$PSCommandPath'"
    | Write-Warning
    # dump json logs
    $publicToExport
    | ConvertTo-Json -Depth 9 -EnumsAsStrings #-EscapeHandling Default
    | Set-Content -Path $__exportPaths['Export_moduleExports_FromList'] -Encoding utf8

    $ninModuleInfo
    | ConvertTo-Json -Depth 9 -EnumsAsStrings #-EscapeHandling Default
    | Set-Content -Path $__exportPaths['Export_moduleExports_FromClass'] -Encoding utf8

    $ninModuleInfo.ExportInfo
    | ConvertTo-Json -Depth 9 -EnumsAsStrings #-EscapeHandling Default
    | Set-Content -Path $__exportPaths['Export_moduleExports_FromClass_Expand'] -Encoding utf8

    $parseResult
    | ConvertTo-Json -Depth 4 -EnumsAsStrings #-EscapeHandling Default
    | Set-Content -Path $__exportPaths['Export_moduleExports_parseResult'] -Encoding utf8

    $parseResult
    | ConvertTo-Json -ea ignore -Depth 4 -EnumsAsStrings #-EscapeHandling Default
    | Set-Content -Path $__exportPaths['Export_moduleExports_parseResult_Summary'] -Encoding utf8

    @(
        $__exportPaths['Export_moduleExports_FromList']
        $__exportPaths['Export_moduleExports_FromClass']
        $__exportPaths['Export_moduleExports_FromClass_Expand']
        $__exportPaths['Export_moduleExports_parseResult']
        $__exportPaths['Export_moduleExports_parseResult_Summary']
    ) | Join-String @strUL
    | Join-String -op "Wrote files`n"
    | Write-Warning

    Write-Warning 'finish formatter, replaces all file info with strings, datetime with string, process as string, etc'
}


if ($____debug.Export_Module_Listing_On_Startup) {
    "`$____debug.Export_Module_Listing_On_Startup: `$true from '$PSCommandPath'"
    # $destPrefix = $PSScriptRoot
    $destPrefix = Get-Item 'C:\nin_temp\1234'

    Write-Verbose "Exporting Modules to: '${destPrefix}/*' "
    Write-Information "Exporting Modules to: '${destPrefix}/*' "

    $parseResultSummary
    | ConvertTo-Json -Depth 20
    | Set-Content -Path (Join-String $destPrefix 'parseResultSummary.json') -Encoding utf8

    $NinModuleInfo
    | ConvertTo-Json -Depth 20
    | Set-Content -Path (Join-String $destPrefix 'ninModuleInfo.json') -Encoding utf8

    $NinModuleInfo.ExportInfo
    | ConvertTo-Json -Depth 20
    | Set-Content -Path (Join-String $destPrefix 'ninModuleInfo-exportInfo.json') -Encoding utf8
    # todo: shared clean and required conditions 2022-03-30
}

# $fileList | ForEach-Object {
#     $RelativePath = "$_.ps1"
#     $src = Join-Path $PSSCriptRoot $RelativePath
#     if (!(Test-Path $src)) {
#         Write-Error  "Unknown import: '$RelativePath'"
#         return
#     }
#     try {
#         . $src

#     }
# catch {
# One nuance of $PSCmdlet.ThrowTerminatingError() is that it creates a terminating error within your Cmdlet but it turns into a non-terminating error after it leaves your Cmdlet.
# but the pipeline ends regardless of a '-Ea Continue'
# $PSCmdlet.ThrowTerminatingError($PSItem)
# $_.InvocationInfo, and inner exception: <https://powershellexplained.com/2017-04-10-Powershell-exceptions-everything-you-ever-wanted-to-know/#psiteminvocationinfo>
# $PSCmdlet.ThrowTerminatingError($PSItem)
# Write-Error "Import failed '$RelativePath'" -Exception $_
# }
# }
# }
# catch {
# Write-Error -Exception $_ -Message 'autoloader failed'
# return
# throw -e
# throw $_
# }
# Export-ModuleMember -Function $funcList

# . (Join-Path $PSSCriptRoot 'Get-CommandSummary.ps1')
# Export-ModuleMember 'Get-CommandSummary-OldMethod'
